netket/graph/abstract_graph.py

```diff
@@ -46,8 +46,8 @@ def distances(self) -> List[List]:
         raise NotImplementedError
 
     @abc.abstractmethod
-    def automorphisms(self):
-        r"""Symmetry group containing the automorphisms of the graph"""
+    def _automorphisms(self):
+        r"""Lists the automorphisms of the graph"""
```

Comment:
So it should be this (with correct import):
```suggestion
    @deprecated("please use `netket.symmetry.automorphism_group(self)` instead")
    def automorphisms(self):
        r"""Lists the automorphisms of the graph"""
```

Reply:
Yup, plus the covariant change in `automorphism_group()`. We don't even need to deprecate it after all, `DenseSymm` takes this format too, for instance. We should just write docs/tutorials without using this function

Reply:
Bu `automorphisms` returns a `PermutationGroup` as well. This should be a drop-in replacement.

Reply:
See here: https://github.com/netket/netket/blob/5340864cc240d7f5f591a0aa3088b9779bc4b888/netket/graph/graph.py#L95
(all our graphs derive from `NetworkX`, so this should cover all of them in current `master`.

Reply:
Oops, needs fixing, see above, it can't return any `Group` without causing circular imports.

Reply:
Since `PermutationGroup` does not need a graph, can't you split the files so that the file it is defined in does not require anything from the graph modukes?

Reply:
See https://github.com/netket/netket/pull/724#issuecomment-846099526, unless we're happy for the space group stuff to live somewhere else from the rest of `symmetry`, `symmetry/__init__.py` will cause a circular import

Reply:
Why does it depend on Lattice exactly? we can avoid a circular import sometimes.

Reply:
`SpaceGroupBuilder` literally has a `Lattice` member because it needs to check positions, wave vectors, etc. against it all the time. Nothing else in the module depends on any `Graph`

Reply:
ok, and where is `spacegroupbuilòder` used in lattice?

Reply:
See this as an example of how to break those loading circles: [lazy.py](https://github.com/netket/netket/blob/5340864cc240d7f5f591a0aa3088b9779bc4b888/netket/operator/_lazy.py#L15) loads abstract operator, and abstract operator loads [lazy](https://github.com/netket/netket/blob/5340864cc240d7f5f591a0aa3088b9779bc4b888/netket/operator/_abstract_operator.py#L318) and

Reply:
It doesn't use SPaceGroupBuilder, but if it is to use any Group, it will call `symmetry/__init.py__` which would load it:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/attila/NQS/netket/netket/__init__.py", line 41, in <module>
    from . import legacy
  File "/home/attila/NQS/netket/netket/legacy/__init__.py", line 1, in <module>
    from netket import (
  File "/home/attila/NQS/netket/netket/exact.py", line 20, in <module>
    from .operator import AbstractOperator
  File "/home/attila/NQS/netket/netket/operator/__init__.py", line 15, in <module>
    from ._abstract_operator import AbstractOperator
  File "/home/attila/NQS/netket/netket/operator/_abstract_operator.py", line 25, in <module>
    from netket.hilbert import AbstractHilbert
  File "/home/attila/NQS/netket/netket/hilbert/__init__.py", line 16, in <module>
    from .custom_hilbert import CustomHilbert
  File "/home/attila/NQS/netket/netket/hilbert/custom_hilbert.py", line 25, in <module>
    from netket.graph import AbstractGraph
  File "/home/attila/NQS/netket/netket/graph/__init__.py", line 17, in <module>
    from .lattice import Lattice
  File "/home/attila/NQS/netket/netket/graph/lattice.py", line 34, in <module>
    from netket.utils.symmetry import PermutationGroup
  File "/home/attila/NQS/netket/netket/utils/symmetry/__init__.py", line 19, in <module>
    from .space_group import SpaceGroupBuilder, space_group
  File "/home/attila/NQS/netket/netket/utils/symmetry/space_group.py", line 30, in <module>
    from netket.graph import Lattice
ImportError: cannot import name 'Lattice' from partially initialized module 'netket.graph' (most likely due to a circular import) (/home/attila/NQS/netket/netket/graph/__init__.py)
```

Reply:
That's why I'm saying that we can put SpaceGroupBuilder somewhere else and be done with it, but that doesn't sound like a great solution either...
And at the end of the day, it's not unreasonable that `netket.symmetry` should deal with anything to do with the symmetries of systems, rather than having a single such feature left behind in `netket.graph` for historical reasons.

Reply:
> See this as an example of how to break those loading circles: [lazy.py](https://github.com/netket/netket/blob/5340864cc240d7f5f591a0aa3088b9779bc4b888/netket/operator/_lazy.py#L15) loads abstract operator, and abstract operator loads [lazy](https://github.com/netket/netket/blob/5340864cc240d7f5f591a0aa3088b9779bc4b888/netket/operator/_abstract_operator.py#L318) and

This is cute. However, I think there is a deeper question here: Should the symmetry groups of graphs be generated in the module for graphs or groups? Right now it's done a bit in both, which will of course lead to bad design and errors like this. It's not like we can't avoid that circular import